nova/scheduler/filters/shard_filter.py

```diff
@@ -114,11 +120,72 @@ def _get_shards(self, project_id):
 
         return self._PROJECT_SHARD_CACHE.get(project_id)
 
-    def host_passes(self, host_state, spec_obj):
+    def _get_k8s_cluster_instances(self, spec_obj):
+        """If the instance will be part of a K8S cluster, it returns
+        the list of all other instances that are already part of it,
+        if any.
+        """
+        # BigVMs are scheduled based on their own rules.
+        if nova_utils.is_big_vm(spec_obj.memory_mb, spec_obj.flavor):
+            return []
+
+        k8s_filter = self._k8s_instance_query_filter(spec_obj)
+
+        if not k8s_filter:
+            return []
+
+        k8s_filter['project_id'] = spec_obj.project_id
+
+        instances = InstanceList.get_by_filters(
+            nova_context.get_admin_context(), filters=k8s_filter,
+            expected_attrs=['flavor', 'metadata', 'tags'])
+
+        if instances and spec_obj.availability_zone:
+            return [i for i in instances
+                    if i.availability_zone == spec_obj.availability_zone]
+
+        return instances
+
+    def _k8s_instance_query_filter(self, spec_obj):
+        elevated = nova_context.get_admin_context()
+        build_request = BuildRequest.get_by_instance_uuid(
+            elevated, spec_obj.instance_uuid)
```

Comment:
If this is a resize, do we we have a `BuildRequest`? We call this function before we "skip for resize", as we pre-compute things.